File: py/colombiapublic_sdk.py
# ColombiaPublic SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import ColombiaPublicUtility
from core.spec import ColombiaPublicSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import ColombiaPublicBaseFeature
from features import _make_feature


class ColombiaPublicSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = ColombiaPublicUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return ColombiaPublicUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = ColombiaPublicSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def airport(self):
        """Idiomatic facade: client.airport.list() / client.airport.load({"id": ...})."""
        from entity.airport_entity import AirportEntity
        cached = getattr(self, "_airport", None)
        if cached is None:
            cached = AirportEntity(self, None)
            self._airport = cached
        return cached

    def Airport(self, data=None):
        # Deprecated: use client.airport instead.
        from entity.airport_entity import AirportEntity
        return AirportEntity(self, data)


    @property
    def category_natural_area(self):
        """Idiomatic facade: client.category_natural_area.list() / client.category_natural_area.load({"id": ...})."""
        from entity.category_natural_area_entity import CategoryNaturalAreaEntity
        cached = getattr(self, "_category_natural_area", None)
        if cached is None:
            cached = CategoryNaturalAreaEntity(self, None)
            self._category_natural_area = cached
        return cached

    def CategoryNaturalArea(self, data=None):
        # Deprecated: use client.category_natural_area instead.
        from entity.category_natural_area_entity import CategoryNaturalAreaEntity
        return CategoryNaturalAreaEntity(self, data)


    @property
    def constitution_article(self):
        """Idiomatic facade: client.constitution_article.list() / client.constitution_article.load({"id": ...})."""
        from entity.constitution_article_entity import ConstitutionArticleEntity
        cached = getattr(self, "_constitution_article", None)
        if cached is None:
            cached = ConstitutionArticleEntity(self, None)
            self._constitution_article = cached
        return cached

    def ConstitutionArticle(self, data=None):
        # Deprecated: use client.constitution_article instead.
        from entity.constitution_article_entity import ConstitutionArticleEntity
        return ConstitutionArticleEntity(self, data)


    @property
    def country(self):
        """Idiomatic facade: client.country.list() / client.country.load({"id": ...})."""
        from entity.country_entity import CountryEntity
        cached = getattr(self, "_country", None)
        if cached is None:
            cached = CountryEntity(self, None)
            self._country = cached
        return cached

    def Country(self, data=None):
        # Deprecated: use client.country instead.
        from entity.country_entity import CountryEntity
        return CountryEntity(self, data)


    @property
    def department(self):
        """Idiomatic facade: client.department.list() / client.department.load({"id": ...})."""
        from entity.department_entity import DepartmentEntity
        cached = getattr(self, "_department", None)
        if cached is None:
            cached = DepartmentEntity(self, None)
            self._department = cached
        return cached

    def Department(self, data=None):
        # Deprecated: use client.department instead.
        from entity.department_entity import DepartmentEntity
        return DepartmentEntity(self, data)


    @property
    def holiday(self):
        """Idiomatic facade: client.holiday.list() / client.holiday.load({"id": ...})."""
        from entity.holiday_entity import HolidayEntity
        cached = getattr(self, "_holiday", None)
        if cached is None:
            cached = HolidayEntity(self, None)
            self._holiday = cached
        return cached

    def Holiday(self, data=None):
        # Deprecated: use client.holiday instead.
        from entity.holiday_entity import HolidayEntity
        return HolidayEntity(self, data)


    @property
    def invasive_specie(self):
        """Idiomatic facade: client.invasive_specie.list() / client.invasive_specie.load({"id": ...})."""
        from entity.invasive_specie_entity import InvasiveSpecieEntity
        cached = getattr(self, "_invasive_specie", None)
        if cached is None:
            cached = InvasiveSpecieEntity(self, None)
            self._invasive_specie = cached
        return cached

    def InvasiveSpecie(self, data=None):
        # Deprecated: use client.invasive_specie instead.
        from entity.invasive_specie_entity import InvasiveSpecieEntity
        return InvasiveSpecieEntity(self, data)


    @property
    def map(self):
        """Idiomatic facade: client.map.list() / client.map.load({"id": ...})."""
        from entity.map_entity import MapEntity
        cached = getattr(self, "_map", None)
        if cached is None:
            cached = MapEntity(self, None)
            self._map = cached
        return cached

    def Map(self, data=None):
        # Deprecated: use client.map instead.
        from entity.map_entity import MapEntity
        return MapEntity(self, data)


    @property
    def native_community(self):
        """Idiomatic facade: client.native_community.list() / client.native_community.load({"id": ...})."""
        from entity.native_community_entity import NativeCommunityEntity
        cached = getattr(self, "_native_community", None)
        if cached is None:
            cached = NativeCommunityEntity(self, None)
            self._native_community = cached
        return cached

    def NativeCommunity(self, data=None):
        # Deprecated: use client.native_community instead.
        from entity.native_community_entity import NativeCommunityEntity
        return NativeCommunityEntity(self, data)


    @property
    def natural_area(self):
        """Idiomatic facade: client.natural_area.list() / client.natural_area.load({"id": ...})."""
        from entity.natural_area_entity import NaturalAreaEntity
        cached = getattr(self, "_natural_area", None)
        if cached is None:
            cached = NaturalAreaEntity(self, None)
            self._natural_area = cached
        return cached

    def NaturalArea(self, data=None):
        # Deprecated: use client.natural_area instead.
        from entity.natural_area_entity import NaturalAreaEntity
        return NaturalAreaEntity(self, data)


    @property
    def president(self):
        """Idiomatic facade: client.president.list() / client.president.load({"id": ...})."""
        from entity.president_entity import PresidentEntity
        cached = getattr(self, "_president", None)
        if cached is None:
            cached = PresidentEntity(self, None)
            self._president = cached
        return cached

    def President(self, data=None):
        # Deprecated: use client.president instead.
        from entity.president_entity import PresidentEntity
        return PresidentEntity(self, data)


    @property
    def radio(self):
        """Idiomatic facade: client.radio.list() / client.radio.load({"id": ...})."""
        from entity.radio_entity import RadioEntity
        cached = getattr(self, "_radio", None)
        if cached is None:
            cached = RadioEntity(self, None)
            self._radio = cached
        return cached

    def Radio(self, data=None):
        # Deprecated: use client.radio instead.
        from entity.radio_entity import RadioEntity
        return RadioEntity(self, data)


    @property
    def region(self):
        """Idiomatic facade: client.region.list() / client.region.load({"id": ...})."""
        from entity.region_entity import RegionEntity
        cached = getattr(self, "_region", None)
        if cached is None:
            cached = RegionEntity(self, None)
            self._region = cached
        return cached

    def Region(self, data=None):
        # Deprecated: use client.region instead.
        from entity.region_entity import RegionEntity
        return RegionEntity(self, data)


    @property
    def touristic_attraction(self):
        """Idiomatic facade: client.touristic_attraction.list() / client.touristic_attraction.load({"id": ...})."""
        from entity.touristic_attraction_entity import TouristicAttractionEntity
        cached = getattr(self, "_touristic_attraction", None)
        if cached is None:
            cached = TouristicAttractionEntity(self, None)
            self._touristic_attraction = cached
        return cached

    def TouristicAttraction(self, data=None):
        # Deprecated: use client.touristic_attraction instead.
        from entity.touristic_attraction_entity import TouristicAttractionEntity
        return TouristicAttractionEntity(self, data)


    @property
    def typical_dish(self):
        """Idiomatic facade: client.typical_dish.list() / client.typical_dish.load({"id": ...})."""
        from entity.typical_dish_entity import TypicalDishEntity
        cached = getattr(self, "_typical_dish", None)
        if cached is None:
            cached = TypicalDishEntity(self, None)
            self._typical_dish = cached
        return cached

    def TypicalDish(self, data=None):
        # Deprecated: use client.typical_dish instead.
        from entity.typical_dish_entity import TypicalDishEntity
        return TypicalDishEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk

File: php/colombiapublic_sdk.php
<?php
declare(strict_types=1);

// ColombiaPublic SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class ColombiaPublicSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new ColombiaPublicUtility();
        $this->_utility = $utility;

        $config = ColombiaPublicConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = ColombiaPublicHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = ColombiaPublicHelpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, ColombiaPublicFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return ColombiaPublicUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = ColombiaPublicHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = ColombiaPublicHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = ColombiaPublicHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new ColombiaPublicSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = ColombiaPublicHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = ColombiaPublicHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_airport = null;

    // Idiomatic facade: $client->airport()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Airport() (PHP method
    // names are case-insensitive).
    public function airport($data = null)
    {
        require_once __DIR__ . '/entity/airport_entity.php';
        if ($data === null) {
            if ($this->_airport === null) {
                $this->_airport = new AirportEntity($this, null);
            }
            return $this->_airport;
        }
        return new AirportEntity($this, $data);
    }


    private $_category_natural_area = null;

    // Idiomatic facade: $client->category_natural_area()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias CategoryNaturalArea() (PHP method
    // names are case-insensitive).
    public function category_natural_area($data = null)
    {
        require_once __DIR__ . '/entity/category_natural_area_entity.php';
        if ($data === null) {
            if ($this->_category_natural_area === null) {
                $this->_category_natural_area = new CategoryNaturalAreaEntity($this, null);
            }
            return $this->_category_natural_area;
        }
        return new CategoryNaturalAreaEntity($this, $data);
    }


    private $_constitution_article = null;

    // Idiomatic facade: $client->constitution_article()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias ConstitutionArticle() (PHP method
    // names are case-insensitive).
    public function constitution_article($data = null)
    {
        require_once __DIR__ . '/entity/constitution_article_entity.php';
        if ($data === null) {
            if ($this->_constitution_article === null) {
                $this->_constitution_article = new ConstitutionArticleEntity($this, null);
            }
            return $this->_constitution_article;
        }
        return new ConstitutionArticleEntity($this, $data);
    }


    private $_country = null;

    // Idiomatic facade: $client->country()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Country() (PHP method
    // names are case-insensitive).
    public function country($data = null)
    {
        require_once __DIR__ . '/entity/country_entity.php';
        if ($data === null) {
            if ($this->_country === null) {
                $this->_country = new CountryEntity($this, null);
            }
            return $this->_country;
        }
        return new CountryEntity($this, $data);
    }


    private $_department = null;

    // Idiomatic facade: $client->department()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Department() (PHP method
    // names are case-insensitive).
    public function department($data = null)
    {
        require_once __DIR__ . '/entity/department_entity.php';
        if ($data === null) {
            if ($this->_department === null) {
                $this->_department = new DepartmentEntity($this, null);
            }
            return $this->_department;
        }
        return new DepartmentEntity($this, $data);
    }


    private $_holiday = null;

    // Idiomatic facade: $client->holiday()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Holiday() (PHP method
    // names are case-insensitive).
    public function holiday($data = null)
    {
        require_once __DIR__ . '/entity/holiday_entity.php';
        if ($data === null) {
            if ($this->_holiday === null) {
                $this->_holiday = new HolidayEntity($this, null);
            }
            return $this->_holiday;
        }
        return new HolidayEntity($this, $data);
    }


    private $_invasive_specie = null;

    // Idiomatic facade: $client->invasive_specie()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias InvasiveSpecie() (PHP method
    // names are case-insensitive).
    public function invasive_specie($data = null)
    {
        require_once __DIR__ . '/entity/invasive_specie_entity.php';
        if ($data === null) {
            if ($this->_invasive_specie === null) {
                $this->_invasive_specie = new InvasiveSpecieEntity($this, null);
            }
            return $this->_invasive_specie;
        }
        return new InvasiveSpecieEntity($this, $data);
    }


    private $_map = null;

    // Idiomatic facade: $client->map()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Map() (PHP method
    // names are case-insensitive).
    public function map($data = null)
    {
        require_once __DIR__ . '/entity/map_entity.php';
        if ($data === null) {
            if ($this->_map === null) {
                $this->_map = new MapEntity($this, null);
            }
            return $this->_map;
        }
        return new MapEntity($this, $data);
    }


    private $_native_community = null;

    // Idiomatic facade: $client->native_community()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias NativeCommunity() (PHP method
    // names are case-insensitive).
    public function native_community($data = null)
    {
        require_once __DIR__ . '/entity/native_community_entity.php';
        if ($data === null) {
            if ($this->_native_community === null) {
                $this->_native_community = new NativeCommunityEntity($this, null);
            }
            return $this->_native_community;
        }
        return new NativeCommunityEntity($this, $data);
    }


    private $_natural_area = null;

    // Idiomatic facade: $client->natural_area()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias NaturalArea() (PHP method
    // names are case-insensitive).
    public function natural_area($data = null)
    {
        require_once __DIR__ . '/entity/natural_area_entity.php';
        if ($data === null) {
            if ($this->_natural_area === null) {
                $this->_natural_area = new NaturalAreaEntity($this, null);
            }
            return $this->_natural_area;
        }
        return new NaturalAreaEntity($this, $data);
    }


    private $_president = null;

    // Idiomatic facade: $client->president()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias President() (PHP method
    // names are case-insensitive).
    public function president($data = null)
    {
        require_once __DIR__ . '/entity/president_entity.php';
        if ($data === null) {
            if ($this->_president === null) {
                $this->_president = new PresidentEntity($this, null);
            }
            return $this->_president;
        }
        return new PresidentEntity($this, $data);
    }


    private $_radio = null;

    // Idiomatic facade: $client->radio()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Radio() (PHP method
    // names are case-insensitive).
    public function radio($data = null)
    {
        require_once __DIR__ . '/entity/radio_entity.php';
        if ($data === null) {
            if ($this->_radio === null) {
                $this->_radio = new RadioEntity($this, null);
            }
            return $this->_radio;
        }
        return new RadioEntity($this, $data);
    }


    private $_region = null;

    // Idiomatic facade: $client->region()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Region() (PHP method
    // names are case-insensitive).
    public function region($data = null)
    {
        require_once __DIR__ . '/entity/region_entity.php';
        if ($data === null) {
            if ($this->_region === null) {
                $this->_region = new RegionEntity($this, null);
            }
            return $this->_region;
        }
        return new RegionEntity($this, $data);
    }


    private $_touristic_attraction = null;

    // Idiomatic facade: $client->touristic_attraction()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias TouristicAttraction() (PHP method
    // names are case-insensitive).
    public function touristic_attraction($data = null)
    {
        require_once __DIR__ . '/entity/touristic_attraction_entity.php';
        if ($data === null) {
            if ($this->_touristic_attraction === null) {
                $this->_touristic_attraction = new TouristicAttractionEntity($this, null);
            }
            return $this->_touristic_attraction;
        }
        return new TouristicAttractionEntity($this, $data);
    }


    private $_typical_dish = null;

    // Idiomatic facade: $client->typical_dish()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias TypicalDish() (PHP method
    // names are case-insensitive).
    public function typical_dish($data = null)
    {
        require_once __DIR__ . '/entity/typical_dish_entity.php';
        if ($data === null) {
            if ($this->_typical_dish === null) {
                $this->_typical_dish = new TypicalDishEntity($this, null);
            }
            return $this->_typical_dish;
        }
        return new TypicalDishEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new ColombiaPublicSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}
